packages/editor/src/components/SvgUploader.tsx

```diff
@@ -0,0 +1,86 @@
+import { FileUploader } from "react-drag-drop-files";
+import toast from "react-hot-toast";
+import { useRecoilState } from "recoil";
+import { fileContentsSelector } from "../state/atoms";
+
+export default function SvgUploader() {
+  const setDomain = useRecoilState(fileContentsSelector("domain"))[1];
+  const setSubstance = useRecoilState(fileContentsSelector("substance"))[1];
+  const setStyle = useRecoilState(fileContentsSelector("style"))[1];
+
+  const handleChange = (svg: File) => {
+    const reader = new FileReader();
+    reader.readAsText(svg);
+    reader.onabort = () => console.log("file reading was aborted");
+    reader.onerror = () => console.log("file reading has failed");
+    reader.onload = () => {
+      const svgText = reader.result?.toString();
+      if (!svgText) {
+        return;
+      }
+
+      const parser = new DOMParser();
+      const xmlDoc = parser.parseFromString(svgText, "text/xml");
+      const styElem = xmlDoc.getElementsByTagName("sty");
+      const subElem = xmlDoc.getElementsByTagName("sub");
+      const dslElem = xmlDoc.getElementsByTagName("dsl");
+
+      if (styElem.length === 0) {
+        toast.error(
+          "Could not load SVG. Make sure the SVG was exported from Penrose."
+        );
+        return;
+      }
+
+      setStyle({
+        name: "SVG import",
+        contents: styElem[0].textContent ?? "".trim(),
+      });
+
+      if (subElem.length === 0) {
+        toast.error(
+          "Could not load SVG. Make sure the SVG was exported from Penrose."
+        );
+        return;
+      }
+
+      setSubstance({
+        name: "SVG import",
+        contents: subElem[0].textContent ?? "".trim(),
+      });
+
+      if (dslElem.length === 0) {
+        toast.error(
+          "Could not load SVG. Make sure the SVG was exported from Penrose."
+        );
+        return;
+      }
+
+      setDomain({
+        name: "SVG import",
+        contents: dslElem[0].textContent ?? "".trim(),
+      });
+
+      toast.success("Sucessfully uploaded SVG to editor");
```

Comment:
Since the SVG should be renderable, we can directly call `compileDiagram` after upload. Can you try using `useCompileDiagram` from `state/callbacks`?

Reply:
bumping this comment